### PR TITLE
Support indexing of the underlying tensors for nested tensors

### LINKED ISSUE #76843: Support indexing of the underlying tensors for nested tensors
### 🚀 The feature, motivation and pitch

Nested tensors were recently added to PyTorch core, but operator support is still fairly sparse. Notably, it's not yet possible to index the underlying tensors:

```python
import torch

x = torch.nested_tensor((torch.randn(3, 4), torch.randn(3, 5)))
print(x[0])  # RuntimeError: Internal error: NestedTensorImpl doesn't support sizes. Please file an issue on https://github.com/pytorch/nestedtensor
```

It'd be great to have simple indexing work to get the underlying tensors from a nested tensor. To start, this could support only indexing on the implicit "batch dimension" that nested tensor provides & error for anything else.

### Alternatives

None

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by YifanShenSZ**
Fixes #76843

### CODE REVIEW

File: aten/src/ATen/TensorIndexing.h [hunk 1]
```diff
@@ -198,16 +198,22 @@ static inline Tensor applySlice(
...
   // TODO: implement negative step
   TORCH_CHECK_VALUE(step > 0, "step must be greater than zero");
 
-  // Skip this optimization if we are tracing, as the trace may be polymorphic
-  // over the shape of the `self` tensor, and we still want to record
-  // the slice.
-  int64_t length = (self_device == at::kCPU || self_device == at::kCUDA) ? self_sizes[dim] : self.size(dim);
-  if (!disable_slice_optimization && start == 0 && stop == length && step == 1) {
-    return self;
+  // TODO: nested tensor does not have a size (yet) so for now we represent its size as null
```
Comment (albanD) [inline aten/src/ATen/TensorIndexing.h:205]: No need to duplicate this comment I think.
we only want a single copy of it at the top level.

File: aten/src/ATen/TensorIndexing.h [hunk 2]
```diff
@@ -431,17 +443,27 @@ static inline Tensor applySlicing(
...
+  if (self_sizes.has_value()) {
+    TORCH_CHECK_INDEX(
+      specified_dims <= (int64_t)self_sizes->size(),
+      "too many indices for tensor of dimension ", (int)self_sizes->size());
+  }
 
   Tensor result = self;
   for (const auto i : c10::irange(indices.size())) {
     auto& obj = indices[i];
+    // TODO: nested tensor does not have a size (yet) so for now we represent its size as null
```
Comment (albanD) [inline aten/src/ATen/TensorIndexing.h:463]: This one is the only one we want to keep.

**Review by albanD — COMMENTED** [review]
Changes to indexing look good (beyond comments).
I'll let Christian comment on the actual implementation

File: aten/src/ATen/TensorIndexing.h [hunk 3]
```diff
@@ -385,7 +397,7 @@ static inline Tensor handleDimInMultiDimIndexing(
       index.slice().step(),
       /*disable_slice_optimization=*/disable_slice_optimization,
       original_tensor_device,
-      prev_dim_result_sizes);
+      *prev_dim_result_sizes);
```
Comment (cpuhrsch) [inline aten/src/ATen/TensorIndexing.h:400]: I think it might be worthwhile adding a `TORCH_INTERNAL_ASSERT_DEBUG_ONLY(prev_dim_result_sizes)` in case this assumption ever does not hold. I'd also annotate that assert with some comment that explains why we can always assume `prev_dim_result_sizes` has a value in this case.

File: aten/src/ATen/TensorIndexing.h [hunk 4]
```diff
@@ -431,17 +443,27 @@ static inline Tensor applySlicing(
...
+      "too many indices for tensor of dimension ", (int)self_sizes->size());
+  }
 
   Tensor result = self;
   for (const auto i : c10::irange(indices.size())) {
     auto& obj = indices[i];
+    // TODO: nested tensor does not have a size (yet) so for now we represent its size as null
+    //       may need to be changed after we reach a better solution for nested tensor size
+    c10::optional<IntArrayRef> result_sizes = c10::nullopt;
+    if (! result.is_nested())  result_sizes = result.sizes();
```
Comment (cpuhrsch) [inline aten/src/ATen/TensorIndexing.h:466]: nit: newline? Your linter should be able to catch things like this.

File: aten/src/ATen/TensorIndexing.h [hunk 5]
```diff
@@ -495,7 +517,10 @@ static inline Tensor dispatch_index_put_(Tensor& self, std::vector<Tensor>&& ind
 // See NOTE [ Setting `disable_slice_optimization` when calling C++ tensor indexing functions from Python ]
 static inline Tensor get_item(const Tensor& self, const ArrayRef<TensorIndex>& indices, bool disable_slice_optimization = false) {
   at::Device self_device = self.device();
-  IntArrayRef self_sizes = self.sizes();
+  // TODO: nested tensor does not have a size (yet) so for now we represent its size as null
+  //       may need to be changed after we reach a better solution for nested tensor size
+  c10::optional<IntArrayRef> self_sizes = c10::nullopt;
+  if (! self.is_nested()) self_sizes = self.sizes();
```
Comment (cpuhrsch) [inline aten/src/ATen/TensorIndexing.h:523]: I think we use curly brackets  even for one 1-line statements. Alternatively I think you can also use a ternary operator (which some people might even like less).
`self.is_nested() ? c10::nullopt : self.sizes()`

File: aten/src/ATen/native/nested/NestedTensorMath.cpp [hunk 6]
```diff
@@ -557,5 +557,39 @@ Tensor& NestedTensor_mul__Tensor(Tensor& self, const Tensor& other) {
...
+  //     * 0, when the nested tensor is empty
+  Tensor sizemat = get_nested_size_tensor(self);
+  TORCH_CHECK(
+    sizemat.dim() == 2,
+    "cannot index an empty nested tensor."
+  );
+  // nested_tensor[i] = i-th original tensor
+  int64_t ntensors = sizemat.size(0);
+  TORCH_CHECK_INDEX(
+    index >= -ntensors && index < ntensors,
```
Comment (cpuhrsch) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:577]: You could use [`maybe_wrap_dim`](https://github.com/pytorch/pytorch/blob/fd6991e71436b26978011cbd49b5d40d27db0857/aten/src/ATen/WrapDimUtils.h) for access to more canonical error messages.

File: aten/src/ATen/native/nested/NestedTensorMath.cpp [hunk 7]
```diff
@@ -557,5 +557,39 @@ Tensor& NestedTensor_mul__Tensor(Tensor& self, const Tensor& other) {
...
+  // size_matrix.dim() can only have 2 possible values:
+  //     * 2, as expected
+  //     * 0, when the nested tensor is empty
+  Tensor sizemat = get_nested_size_tensor(self);
+  TORCH_CHECK(
+    sizemat.dim() == 2,
+    "cannot index an empty nested tensor."
+  );
+  // nested_tensor[i] = i-th original tensor
+  int64_t ntensors = sizemat.size(0);
```
Comment (cpuhrsch) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:575]: Until we add support for [`size.int`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L4284-L4288) to NestedTensor itself (note that this is different from `sizes`) you can also use [`opt_size`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/NestedTensorImpl.h#L25-L31) to get this information. 

As an aside, a convenient way of defining `size.int` is as
```
int64_t size(const Tensor& self, int64_t dim) {
    return *get_nested_tensor_impl(self)->opt_sizes(dim)
}
```

This will throw an error message if `size` has been requested for an irregular dimension. It'll also allow a user to construct `sizes` *if all dimensions are regular* at which point a NestedTensor duck types to become a regular Tensor.

File: aten/src/ATen/native/nested/NestedTensorMath.cpp [hunk 8]
```diff
@@ -557,5 +557,39 @@ Tensor& NestedTensor_mul__Tensor(Tensor& self, const Tensor& other) {
...
+  TORCH_CHECK_INDEX(
+    index >= -ntensors && index < ntensors,
+    "index ", index, " is out of bounds for nested tensor dimension 0 with size ", ntensors);
+  int64_t positive_index = index < 0 ? index + ntensors : index;
+  // buffer contains the underlying data in a contiguous vector
+  Tensor buffer = get_buffer(self);
+  // determine the memory section of the i-th original tensor
```
Comment (cpuhrsch) [inline aten/src/ATen/native/nested/NestedTensorMath.cpp:582]: Nit: The calculation you have in mind here, converting from a nested size to a list of offsets, conceptually already exists in [NestedTensorTransformerFunctions.cpp](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/nested/NestedTensorTransformerFunctions.cpp#L88-L105). Without wanting to push for premature optimization (i.e. no need to address this) note that it is much more efficient to operate on the nested size directly due to overhead from calling `at::prod` and the fact that sizemat is rather small.

Comment (cpuhrsch) [reply aten/src/ATen/native/nested/NestedTensorMath.cpp:582]: It'd also be a useful utility for unbind [above in this file](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/nested/NestedTensorMath.cpp#L113-L143).

File: test/test_nestedtensor.py [hunk 9]
```diff
@@ -355,6 +355,54 @@ def test_device_checks(self, device):
...
+        self.assertRaises(NotImplementedError, lambda: nt[None])
+        self.assertRaises(NotImplementedError, lambda: nt[...])
+        # tuple of indices: only support integer in the batch dimension
+        #                 + all possible indexing in the original tensor dimensions
+        self.assertEqual(nt[0, 0, 0], x0[0, 0])
+        self.assertEqual(nt[0, 1, :], x0[1, :])
+        self.assertEqual(nt[1, ...], x1)
+        self.assertRaisesRegex(
+            IndexError,
+            "index 4 is out of bounds for dimension 1 with size 3",
```
Comment (cpuhrsch) [inline test/test_nestedtensor.py:394]: nit: Technically this error message is incorrect, since dimension 1 of this particular NestedTensor doesn't have a size (it's irregular). 

```
x0 = torch.randn((2, 5), device=device)
x1 = torch.randn((3, 4), device=device)
nt = torch.nested_tensor([x0, x1])
```
has nested size
```
[ 
  (2, 5),
  (3, 4)
]
```

File: test/test_nestedtensor.py [hunk 10]
```diff
@@ -355,6 +355,54 @@ def test_device_checks(self, device):
...
+            lambda: nt[1, 4, 2]
+        )
+        self.assertRaises(NotImplementedError, lambda: nt[:, 1, 1])
+        # make sure indexing returns a view
```
Comment (cpuhrsch) [inline test/test_nestedtensor.py:398]: @albanD @drisspg - this will become more interesting once NestedTensor has autograd support.

File: aten/src/ATen/TensorIndexing.h (hunk 5 above)
Comment (YifanShenSZ) [reply aten/src/ATen/TensorIndexing.h:523]: yeah I tried but I would get a compilation error:
```
../aten/src/ATen/TensorIndexing.h:453:66: error: incompatible operand types ('const c10::nullopt_t' and 'c10::IntArrayRef' (aka 'ArrayRef<long long>'))
    c10::optional<IntArrayRef> result_sizes = result.is_nested() ? c10::nullopt : result.sizes();
```
could be solved though, I think, by constructing `c10::optional<IntArrayRef>` instances from `c10::nullopt` and `result.sizes()`

File: torch/csrc/autograd/python_variable_indexing.cpp [hunk 11]
```diff
@@ -137,18 +137,28 @@ static inline Variable applySlicing(
...
     const at::Device& self_device,
-    const IntArrayRef& self_sizes,
+    const c10::optional<IntArrayRef> & self_sizes,
     int64_t specified_dims) {
   int64_t size = PyTuple_GET_SIZE(index); // NOLINT(cppcoreguidelines-pro-type-cstyle-cast)
   int64_t dim = 0;
 
-  if (specified_dims > (int64_t)self_sizes.size()) {
-    throw IndexError("too many indices for tensor of dimension %d", (int)(self_sizes.size()));
+  // TODO: nested tensor does not have a size (yet) so for now we represent its size as null
```
Comment (cpuhrsch) [inline torch/csrc/autograd/python_variable_indexing.cpp:145]: For repeat notes like this use a "tagging" style and write the note in a single location. For example https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/TensorIndexing.h#L365
```
// See NOTE [ Setting `disable_slice_optimization` when calling C++ tensor indexing functions from Python ]
static inline Tensor handleDimInMultiDimIndexing(
```

which is then defined in a single location https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/TensorIndexing.h#L469-L495

**Comment by cpuhrsch** [comment]
Generally looks great! Thank you for sending this!

**Comment by cpuhrsch** [comment]
@YifanShenSZ - could you also try to rebase or merge master please? There is a conflict that seems to prevent CI from running, which can hide lint failures and the like.

File: test/test_nestedtensor.py [hunk 12]
```diff
@@ -355,6 +355,54 @@ def test_device_checks(self, device):
...
+        nt[0].fill_(100.0)
+        answer = torch.randn((2, 5), device=device).fill_(100.0)
```
Comment (jbschlosser) [inline test/test_nestedtensor.py:400]:
```suggestion
        answer = torch.tensor(100., device=device).expand(2, 5)
```

File: test/test_nestedtensor.py [hunk 13]
```diff
@@ -355,6 +355,54 @@ def test_device_checks(self, device):
...
+        self.assertEqual(nt[0], answer)
+        nt[1, 1, :].fill_(200.0)
+        answer = torch.randn(4, device=device).fill_(200.0)
```
Comment (jbschlosser) [inline test/test_nestedtensor.py:403]:
```suggestion
        answer = torch.tensor(200., device=device).expand(4)
```

File: test/test_nestedtensor.py [hunk 14]
```diff
@@ -355,6 +355,54 @@ def test_device_checks(self, device):
         is_cuda = 'cuda' in str(device)
         self.assertEqual(nt.is_cuda, is_cuda)
 
+    @dtypes(torch.float, torch.float16, torch.double)
+    def test_nested_tensor_indexing(self, device, dtype):
```
Comment (jbschlosser) [inline test/test_nestedtensor.py:359]: Test coverage looks pretty solid!

File: test/test_nestedtensor.py [hunk 15]
```diff
@@ -355,6 +355,54 @@ def test_device_checks(self, device):
...
+        # edge case: empty nested tensor
+        nt0 = torch.nested_tensor([])
+        self.assertRaisesRegex(
+            RuntimeError,
+            "cannot index an empty nested tensor",
+            lambda: nt0[0]
+        )
+        # normal case
+        x0 = torch.randn((2, 5), device=device)
```
Comment (jbschlosser) [inline test/test_nestedtensor.py:368]: Need `dtype=dtype` for tensor creation throughout to test across the dtypes

**Review by jbschlosser — COMMENTED** [review]
Thanks for the update! Added some minor testing comments - implementation looks great

File: aten/src/ATen/TensorIndexing.h (hunk 3 above)
Comment (YifanShenSZ) [reply aten/src/ATen/TensorIndexing.h:394]: It actually can be empty: when using a tuple to index a nested tensor, will call `applySlicing` -> repeat `handleDimInMultiDimIndexing` for all indices, which means at the first pass it can be handling a nested tensor who has empty `prev_dim_result_sizes`

Comment (YifanShenSZ) [reply aten/src/ATen/TensorIndexing.h:394]: but for latter passes it must have a value, though, since it would always be normal tensor by then

File: test/test_nestedtensor.py (hunk 9 above)
Comment (YifanShenSZ) [reply test/test_nestedtensor.py:394]: Yeah that's an issue. In 2nd and latter `handleDimInMultiDimIndexing` calls, it works on the original normal tensor, which means the error messages become the ones for the original normal tensor. I can't think of a way to get rid of it

File: test/test_nestedtensor.py (hunk 10 above)
Comment (YifanShenSZ) [reply test/test_nestedtensor.py:398]: I considered nested tensor slicing, but the best way I could come up with is to only support slice 1: the sizes of the original tensors differ, so the stride can be irregular. Decided to leave it until we reach a better solution

**Comment by YifanShenSZ** [comment]
> @YifanShenSZ - could you also try to rebase or merge master please? There is a conflict that seems to prevent CI from running, which can hide lint failures and the like.

just merged master. CI is running now

File: test/test_nestedtensor.py (hunk 10 above)
Comment (cpuhrsch) [reply test/test_nestedtensor.py:398]: It is possible to support slicing and return a NestedTensor. It's a worthwhile extension, but I don't think we need it in the context of the PR.

File: test/test_nestedtensor.py (hunk 9 above)
Comment (cpuhrsch) [reply test/test_nestedtensor.py:394]: Makes sense. I'd note that as a follow-up item and create a small, separate issue to track it.

File: aten/src/ATen/TensorIndexing.h (hunk 3 above)
Comment (cpuhrsch) [reply aten/src/ATen/TensorIndexing.h:394]: So does that mean that adding this check would be redundant or that `prev_dim_result_sizes` always has a value, but might be an empty IntArrayRef? If it always has a value, does this need to be an optional IntArrayRef or can it just be a regular IntArrayRef that happens to be empty sometimes?

Comment (YifanShenSZ) [reply aten/src/ATen/TensorIndexing.h:394]: I should be more precise: `prev_dim_result_sizes` is `c10::nullopt` in the first call, then regular `IntArrayRef` in the latter calls

Comment (YifanShenSZ) [reply aten/src/ATen/TensorIndexing.h:394]: so we cannot have the `TORCH_INTERNAL_ASSERT_DEBUG_ONLY(prev_dim_result_sizes)`: it does not hold in the first call

Comment (cpuhrsch) [reply aten/src/ATen/TensorIndexing.h:394]: But it has to hold true in this particular branch, otherwise `* prev_dim_result_sizes` should [violate an assertion](https://github.com/pytorch/pytorch/blob/master/c10/util/Optional.h#L739-L751).

Comment (jbschlosser) [reply aten/src/ATen/TensorIndexing.h]: @cpuhrsch the updated version of the code no longer unconditionally dereferences `prev_dim_result_sizes` for this branch

**Review by cpuhrsch — APPROVED** [review]
LGTM from a NestedTensor perspective

**Review by jbschlosser — APPROVED** [review]
LGTM
